README.md
# Read All Literature

This is a mod for the Project Zomboid that adds two options to containers with books. You can either collect all the books that you have not yet acquired in your inventory, or have your character read all the books in the container that they don't know yet, one after the other.

## Release Notes

You can find the latest release notes [here](https://github.com/Amadon/ReadAllLiterature/blob/main/release-notes.md).

## Credits

- Thanks to tyriell for the idea
- Thanks to PePePePeil and his mod "Has Been Read" for helping to understand the Lua code (https://steamcommunity.com/sharedfiles/filedetails/?id=2544353492)

## Installation

The mod can be installed through the Steam Workshop with the ID: 2898147094

## Mod ID

ReadAllLiterature


Changelog.md
# Changelog

## V1.1

- Individually collected books are now also included in the collection
- Translations for English and German added
- Code revision and clean layout
- Books are put back after reading

## Known Bugs

- Magazines cause an error in the queue when they are put back. Currently, only talent books are put back due to the area being commented out.

## Future Plans

- Bugfixes
- Option to make all features optional in the server configuration
- Own bookshelf with enough space for all books and magazines
- Swarm Intelligence to collect books in groups and prevent other players from taking books that are already known in the group
- New books for the game. Wishes and suggestions are welcome in the comments.


Contents/mods/ReadAllLiterature/media/lua/client/ReadAllLiterature/ReadAllLiterature.lua
ReadAllLiterature = {
	readableArray = {}
}

function ReadAllLiterature:isSkillBook(playerObj, item)
	local isSkillBook = false
	local skillBook = SkillBook[item:getSkillTrained()]
	
    if skillBook then
		local perkLevel = playerObj:getPerkLevel(skillBook.perk)
		local minLevel = item:getLvlSkillTrained()
		local maxLevel = item:getMaxLevelTrained()
		if (minLevel <= perkLevel + 1) and (perkLevel + 1 <= maxLevel) then
			local readPages = playerObj:getAlreadyReadPages(item:getFullType())
			if readPages >= item:getNumberOfPages() then
				isSkillBook = false
			elseif perkLevel >= maxLevel then
				isSkillBook = false
			elseif readPages > 0 then
				isSkillBook = true
			else
				isSkillBook = true
			end
		else
			isSkillBook = false
		end
	end

	return isSkillBook
end

function ReadAllLiterature:isRecipes(playerObj, item)
	local isRecipes = false

	if item:getTeachedRecipes() and not item:getTeachedRecipes():isEmpty() then
		if playerObj:getKnownRecipes():containsAll(item:getTeachedRecipes()) then
			isRecipes = false
		else
			isRecipes = true
		end
	end

	return isRecipes
end

function ReadAllLiterature:isReadable(playerObj, item)
	local isReadable = false
	local isLiterature = item:getCategory() == "Literature"
	if isLiterature then
		local isSkillBook = ReadAllLiterature:isSkillBook(playerObj, item)
		local isRecipes = ReadAllLiterature:isRecipes(playerObj, item)
		if  isSkillBook or isRecipes then
			isReadable = true
		end
	end
	return isReadable
end

function ReadAllLiterature:ReadableFilter(container)
	local playerObj = getPlayer()
	local readableArray = {}
	if container:getItems():size() > 0 then
		for i = 0, container:getItems():size() -1 do
			local item = container:getItems():get(i)
			if ReadAllLiterature:isReadable(playerObj, item) then
				table.insert(readableArray, item)
			end
		end
	end
	return readableArray
end

function ReadAllLiterature:perform()
	local playerObj = getPlayer()
	local magazines = {}
	for _, item in ipairs(ReadAllLiterature.readableArray) do
		if ReadAllLiterature:isReadable(playerObj, item) then
			local srcContainer = item:getContainer()
			if luautils.haveToBeTransfered(playerObj, item) then
				ISTimedActionQueue.add(ISInventoryTransferAction:new(playerObj, item, item:getContainer(), playerObj:getInventory()))
			end
			ISTimedActionQueue.add(ISReadABook:new(playerObj, item, 150))
			if SkillBook[item:getSkillTrained()] then
				ISTimedActionQueue.add(ISInventoryTransferAction:new(playerObj ,item,  playerObj:getInventory(), srcContainer))
			else
				table.insert(magazines, item)
				if #magazines >= 5 then
					ISTimedActionQueue.add(ISInventoryTransferAction:new(playerObj, magazines, playerObj:getInventory(), srcContainer))
					magazines = {}
				end
			end
		end
	end
	if #magazines > 0 then
		ISTimedActionQueue.add(ISInventoryTransferAction:new(playerObj, magazines, playerObj:getInventory(), srcContainer))
	end
end

function ReadAllLiterature:AddReadButtonEvent(state)
	if state == "end" and not self.onCharacter then
		local leftSpace = self.lootAll:getRight() + 16
		ReadAllLiterature.readableArray = ReadAllLiterature:ReadableFilter(self.inventoryPane.inventory)
		if #(ReadAllLiterature.readableArray) > 0 then
			self.readButton = ISButton:new(leftSpace, 0, 50, self:titleBarHeight(), getText("UI_ReadAllLiterature_Read_All"), self, ReadAllLiterature.perform);
			self.readButton:initialise();
			self.readButton.borderColor.a = 0.0;
			self.readButton.backgroundColor.a = 0.0;
			self.readButton.backgroundColorMouseOver.a = 0.7;
			self:addChild(self.readButton);
			self.readButton:setVisible(true);
		end
	elseif state == "begin" and not self.onCharacter and self.readButton then
		self.readButton:setVisible(false);
	end
end


Contents/mods/ReadAllLiterature/media/lua/client/ReadAllLiterature/CollectAllLiterature.lua
CollectAllLiterature = {
	collectableArray = {}
}

-- Get the book list for the player or group,
-- depending on if the player is in a group or not.
function CollectAllLiterature:getBookList()
    local player = getPlayer()  -- Get the player object.
    local currentGroup = player:getGroup()  -- Get the player's current group.
    local modData  -- Declare the modData variable.

    -- Check if the player is in a group.
    if currentGroup then
        modData = currentGroup:getModData()  -- Get the group's modData.
    else
        modData = player:getModData()  -- Get the player's modData.
    end

    -- Check if the CollectAllLiterature field exists in the modData.
    if not modData.CollectAllLiterature then
        modData.CollectAllLiterature = {}  -- Create the CollectAllLiterature field.
        modData.CollectAllLiterature.bookList = {}  -- Create the bookList field.
    end

    return modData.CollectAllLiterature.bookList  -- Return the book list.
end


function CollectAllLiterature:addToBookList(bookItem)
	local isCollectable = CollectAllLiterature:isCollectable(bookItem)
	
	if isCollectable then
		local booksList = CollectAllLiterature:getBookList()
		local fullName = bookItem:getFullType()
		table.insert(booksList, fullName)
	end
	
	return isCollectable
end

function CollectAllLiterature:isInBookList(item)
	local isInBookList = false
	local fullBookName = item:getFullType()
    local bookList = CollectAllLiterature:getBookList()
	if bookList then
		for _, book in pairs(bookList) do
			if book == fullBookName then
				isInBookList = true
			end
		end
	end
	return isInBookList
end

function CollectAllLiterature:isRecipe(item)
	return item:getTeachedRecipes()
end

function CollectAllLiterature:isSkillBook(item)
	return SkillBook[item:getSkillTrained()]
end

--[[
This function checks whether an item is collectable based on several conditions.

The conditions are:
- the item is not already in the collection book list
- the item is either a literature item (i.e., a skill book or a recipe), or a map

Input:
- item: the item to check

Output:
- a boolean indicating whether the item is collectable
]]
function CollectAllLiterature:isCollectable(item)
    -- If the item is already in the collection book list, it is not collectable
    if CollectAllLiterature:isInBookList(item) then
        return false
    end

	-- By default, the item is not considered to be literature
    local isLiterature = false
    -- Check if the item category is "Literature"
    local itemCategory = item:getCategory()
    if itemCategory == "Literature" then
	    -- If the item category is "Literature", check if it is a skill book or a recipe
        local isSkillBook = ModOptions.getOption("Collect", "Books") and CollectAllLiterature:isSkillBook(item)
		local isRecipe =  ModOptions.getOption("Collect", "Recipes") and CollectAllLiterature:isRecipe(item)
        -- If the item is either a skill book or a recipe, it is considered to be literature
        isLiterature = isSkillBook or isRecipe
    end
	-- Check if the item type is "Map"
    local isMap = ModOptions.getOption("Collect", "Maps") and item:getType() == "Map"
    -- Return the result: either the item is literature, or it is a map, or both
    return isLiterature or isMap
end

function CollectAllLiterature:CollectableFilter(container)
	local collectableArray = {}
	if container:getItems():size() > 0 then
		for i = 0, container:getItems():size() -1 do
			local item = container:getItems():get(i)
			if CollectAllLiterature:isCollectable(item) then
				table.insert(collectableArray, item)
			end
		end
	end
	return collectableArray
end

function CollectAllLiterature:perform()
	local playerObj = getPlayer()
	for _, item in ipairs(CollectAllLiterature.collectableArray) do
		if CollectAllLiterature:isCollectable(item) and luautils.haveToBeTransfered(playerObj, item) then
			ISTimedActionQueue.add(ISInventoryTransferAction:new(playerObj, item, item:getContainer(), playerObj:getInventory()))
		end
	end
end

function CollectAllLiterature:AddCollectButtonEvent(state)
	if state == "end" and not self.onCharacter then
		CollectAllLiterature.collectableArray = CollectAllLiterature:CollectableFilter(self.inventoryPane.inventory)
		local leftSpace = self.lootAll:getRight() + 16
		if self.readButton and self.readButton:isVisible() then
			leftSpace = self.readButton:getRight() + 16
		end
		if (CollectAllLiterature.collectableArray) > 0 then
  			self.collectButton = ISButton:new(leftSpace, 0, 50, self:titleBarHeight(), getText("UI_CollectAllLiterature_Collect_All"), self, CollectAllLiterature.perform);
			self.collectButton:initialise();
			self.collectButton.borderColor.a = 0.0;
			self.collectButton.backgroundColor.a = 0.0;
			self.collectButton.backgroundColorMouseOver.a = 0.7;
			self:addChild(self.collectButton);
			self.collectButton:setVisible(true);
		end
	elseif state == "begin" and not self.onCharacter and self.collectButton then
		self.collectButton:setVisible(false);
	end
end